### DISCUSSION TITLE
Overwrite runs-on defined in workflow_call

### DISCUSSION THREAD

**Discussion by nissessenap** [community/community#26801] [Actions]
<p>I’m creating a general workflow_call to create easy to use templates that is supposed to be used by both companies and open-source.<br>
Some users will use self hosted servers and others will use “ubuntu-latest”</p>
<p>To my knowledge I have to define <code>runs-on</code> in my workflow_call action:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://docs.github.com/en/actions/learn-github-actions/reusing-workflows#example-reusable-workflow">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">

      <a href="https://docs.github.com/en/actions/learn-github-actions/reusing-workflows#example-reusable-workflow" target="_blank" rel="noopener nofollow ugc">GitHub Docs</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/14774899/181099952-d18f21be-89e9-4d45-845a-022c312d02e2.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://docs.github.com/en/actions/learn-github-actions/reusing-workflows#example-reusable-workflow" target="_blank" rel="noopener nofollow ugc">Reusing workflows - GitHub Docs</a></h3>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Is there any way to overwrite runs-on so I instead can use<br>
runs-on: [self-hosted, linux] where I want?</p>

**Comment by ChristopherHX** [comment, marked as answer]
<p>I didn’t think about this yet, it seems you can use expression syntax in runs-on with your input.</p>
<p>Example</p>
<pre><code class="lang-yaml">
on:
  workflow_call:
    inputs:
      runs-on:
        type: string
        default: '["windows-latest"]'
jobs:
  myjoba:
    runs-on: ${{fromJSON(inputs.runs-on)}}
    outputs:
      myout: myin
    steps:
    - run: echo Hello World
</code></pre>
<p>Call the workflow with runs-on as input.<br>
Multiple labels</p>
<pre><code class="lang-yaml">on: push
jobs:
  comp:
    name: comp-${{toJSON(strategy)}}
    uses: ChristopherHX/newcomposite-sample/.github/workflows/weirdworkflow.yml@reusable-workflows
    with:
      runs-on: '["self-hosted", "linux"]'
</code></pre>
<p>Single label</p>
<pre><code class="lang-yaml">on: push
jobs:
  comp:
    uses: ChristopherHX/newcomposite-sample/.github/workflows/weirdworkflow.yml@reusable-workflows
    with:
      runs-on: '"windows-latest"'
</code></pre>
<p>It isn’t great that inputs cannot be Objects (Arrays), but only string, boolean or number. You have to use json, unless github allows object inputs in the furture.</p>
<p><em>I’m not shure where inputs are allowed and where it is forbidden, the documentation is outdated so I had to test it myself and it worked</em></p>

**Comment by nissessenap** [comment]
<p>Easier solution then I hoped for, I should have thought of it as well.<br>
Thanks for the quick reply and the great example!</p>

**Comment by jricciardi** [comment]
<p>Thanks for jumping in with that speedy reply, <a class="mention" href="/christopherhx">@ChristopherHX</a>! Glad to hear this helped, <a class="mention" href="/nissessenap">@NissesSenap</a> :slight_smile:</p>
